pyccl/tests/test_power.py

```diff
@@ -187,89 +205,278 @@ def test_power_sigma8norm_norms_consistent(tf):
-                                            hoh0_array=hoh0_from_ccl)
-    cosmo_input._set_growth_from_arrays(a_array=a_arr,
-                                        growth_array=growth_from_ccl,
-                                        fgrowth_array=fgrowth_from_ccl)
-    cosmo_input._set_nonlin_power_from_arrays(a_arr, k_arr, pk_arr)
+    cosmo_input = ccl.CosmologyCalculator(
+        Omega_c=0.27, Omega_b=0.05, h=0.7,
+        n_s=0.965, A_s=2e-9,
+        background={'a': a_arr,
+                    'chi': chi_from_ccl,
+                    'h_over_h0': hoh0_from_ccl},
+        growth={'a': a_arr,
+                'growth_factor': growth_from_ccl,
+                'growth_rate': fgrowth_from_ccl},
+        pk_nonlin={'a': a_arr, 'k': k_arr,
+                   'delta_matter_x_delta_matter': pk_arr})
 
     pk_CCL_input = ccl.power.nonlin_matter_power(cosmo_input, k_arr, 0.5)
     pk_CCL = ccl.power.nonlin_matter_power(cosmo, k_arr, 0.5)
 
     assert np.allclose(pk_CCL_input, pk_CCL, atol=0., rtol=1e-5)
 
+    # Test again with negative power spectrum (so it's not logscaled)
+    cosmo_input = ccl.CosmologyCalculator(
+        Omega_c=0.27, Omega_b=0.05, h=0.7,
+        n_s=0.965, A_s=2e-9,
+        background={'a': a_arr,
+                    'chi': chi_from_ccl,
+                    'h_over_h0': hoh0_from_ccl},
+        growth={'a': a_arr,
+                'growth_factor': growth_from_ccl,
+                'growth_rate': fgrowth_from_ccl},
+        pk_nonlin={'a': a_arr, 'k': k_arr,
+                   'delta_matter_x_delta_matter': -pk_arr})
+
+    pk_CCL_input = -ccl.power.nonlin_matter_power(cosmo_input, k_arr, 0.5)
+
+    assert np.allclose(pk_CCL_input, pk_CCL, atol=0., rtol=1e-5)
+
+    # Test again passing only linear Pk, but letting HALOFIT do its thing
+    kl_arr = np.logspace(-4, 1, 1000)
+    pkl_arr = np.array([ccl.power.linear_matter_power(cosmo, kl_arr, a)
+                        for a in a_arr])
+    cosmo_input = ccl.CosmologyCalculator(
+        Omega_c=0.27, Omega_b=0.05, h=0.7,
+        n_s=0.965, A_s=2e-9,
+        pk_linear={'a': a_arr, 'k': kl_arr,
+                   'delta_matter_x_delta_matter': pkl_arr},
+        use_halofit=True)
+
+    pk_CCL_input = ccl.power.nonlin_matter_power(cosmo_input, k_arr, 0.5)
+
+    assert np.allclose(pk_CCL_input, pk_CCL, atol=0., rtol=1e-5)
+
+    # Test extra power spectrum
+    kl_arr = np.logspace(-4, 1, 1000)
+    pkl_arr = np.array([ccl.power.linear_matter_power(cosmo, kl_arr, a)
+                        for a in a_arr])
+    cosmo_input = ccl.CosmologyCalculator(
+        Omega_c=0.27, Omega_b=0.05, h=0.7,
+        n_s=0.965, A_s=2e-9,
+        pk_linear={'a': a_arr, 'k': kl_arr,
+                   'delta_matter_x_delta_matter': pkl_arr,
+                   'a_x_b': pkl_arr},
+        pk_nonlin={'a': a_arr, 'k': k_arr,
+                   'delta_matter_x_delta_matter': -pk_arr},
+        use_halofit=True)
```

Comment:
This is going to trip people up and create weird bugs where people think they are getting one thing but not another. I wonder if we should not allow this calling signature and make people be more explicit about what they want? We could force use_halofit to be a dict in this case mapping the linear power name to True or False.

Reply:
See my comment/questions below.